src/modalities/models/huggingface_adapters/hf_adapter.py

```diff
@@ -0,0 +1,87 @@
+import json
+from dataclasses import dataclass
+from pathlib import PosixPath
+from typing import Any, Dict, Optional, Tuple
+
+import torch
+import torch.nn as nn
+from transformers import PreTrainedModel, PretrainedConfig
+from transformers.utils import ModelOutput
+
+from modalities.models.model import NNModel
+from modalities.models.utils import get_model_from_config
+
+
+class HFAdapterConfig(PretrainedConfig):
+    model_type = "modalities"
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        if self.config:
+            self.convert_posixpath_to_str(self.config)
+
+    def to_json_string(self, use_diff: bool = True) -> str:
+        if self.config:
+            json_dict = {"config": self.config.copy(), "model_type": self.model_type}
+        else:
+            json_dict = {}
+        return json.dumps(json_dict)
+
+    def convert_posixpath_to_str(self, d: dict):
+        """
+        Recursively iterate over the dictionary and convert PosixPath values to strings.
+        """
+        for key, value in d.items():
+            if isinstance(value, PosixPath):
+                d[key] = str(value)
+            elif isinstance(value, dict):
+                self.convert_posixpath_to_str(value)
+            elif isinstance(value, list):
+                d[key] = [str(item) if isinstance(item, PosixPath) else item for item in value]
+
+
+class HFAdapter(PreTrainedModel):
+    config_class = HFAdapterConfig
+
+    def __init__(self, config: HFAdapterConfig, model: Optional[nn.Module] = None, *inputs, **kwargs):
+        super().__init__(config)
+        if not model:
+            self.model: NNModel = get_model_from_config(config.config)
+        else:
+            self.model = model
+
+    def forward(
+        self,
+        input_ids: torch.Tensor,
+        attention_mask: Optional[torch.Tensor] = None,
+        return_dict: Optional[bool] = False,
+        output_attentions: Optional[bool] = False,
+        output_hidden_states: Optional[bool] = False,
+    ):
+        if output_attentions or output_hidden_states:
+            raise NotImplementedError
+        model_input = {"input_ids": input_ids, "attention_mask": attention_mask}
+        model_forward_output: Dict[str, torch.Tensor] = self.model.forward(model_input)
+        if return_dict:
+            return ModalitiesModelOutput(**model_forward_output)
+        else:
+            return model_forward_output[self.model.prediction_key]
```

Comment:
We think in our case it is necessary that the model has a prediction key. Else how should we know which entry to access in model_forward_output? We add therefore an explicit assertion to ensure that model.prediction_key exists

Reply:
The issue is that `NNModel`does not have the attribute `prediction_key` and cannot assume that every inherited model has it. So by "contract" of the class definitions, it is totally fine that a model does not have a `prediction_key` and it would throw an error in the HFAdapter at runtime. 

Additionally, for different use cases other prediction keys might be expected. E.g., argmax vs softmax outputs are both predictions that we could use in different scenarios. 

I would therefore specify the needed prediction key for the HFAdpater in the config yaml  by referencing the key.